lib/AST/DeclContext.cpp

```diff
@@ -790,29 +813,83 @@ void IterableDeclContext::addMember(Decl *member, Decl *Hint) {
   }
 }
 
-void IterableDeclContext::addMemberSilently(Decl *member, Decl *hint) const {
+void IterableDeclContext::addMemberSilently(Decl *member, Decl *hint,
+                                            bool insertAtHead) const {
   assert(!isa<AccessorDecl>(member) && "Accessors should not be added here");
   assert(!member->NextDecl && "Already added to a container");
 
-  // If there is a hint decl that specifies where to add this, just
-  // link into the chain immediately following it.
-  if (hint) {
+#ifndef NDEBUG
+  auto checkSourceRange = [&](Decl *prev, Decl *next) {
+    if (!member->getDeclContext()->getParentSourceFile())
+      return;
+
+    auto shouldSkip = [](Decl *d) {
+      if (isa<VarDecl>(d) || isa<EnumElementDecl>(d) || isa<IfConfigDecl>(d))
+        return true;
+
+      if (d->isImplicit())
+        return true;
+
+      return false;
+    };
+
+    if (shouldSkip(prev) || shouldSkip(next))
+      return;
+
+    SourceLoc prevEnd = prev->getEndLoc();
+    SourceLoc nextStart = next->getStartLoc();
+
+    if (!prevEnd.isValid() || !nextStart.isValid())
+      return;
+
+    if (getASTContext().SourceMgr.isBeforeInBuffer(prevEnd, nextStart))
+      return;
+
+    llvm::errs() << "Source ranges out of order in addMember():\n";
+    prev->dump(llvm::errs());
+    next->dump(llvm::errs());
+    abort();
+  };
+#endif
```

Comment:
Great to see the assertion here! A couple of thoughts:
1. A lambda inside the function vs a separate function:
  In favor of the lambda is that it is only called here. In favor of a separate function is that `addMemberSilently` would be a bit more readable without the lambda in the middle. Also the data flow into `checkSourceRange` gets clearer when forced to pass in data explicitly. (Obviously I lean towards the latter.)

2. Does it make sense to generalize the order checking to any Decl? The details would be different for brace statements, but it would be nice to reify the commonality of needing source ranges to be ordered.

BTW, great to check the invariant right at the insertion point!

Reply:
Also, I think some "why" comments could help here. Why check the ranges? Why skip over vars, enums, etc.?

Reply:
I'll add some more comments, thanks.

I'll also see about adding a similar assertion to SourceFile::addTopLevelDecl() and BraceStmt::setBody() (and hopefully share code), but I'll do that in a separate PR.

Reply:
Great! Thank you.